app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/cloudCity/NetworkClient.java

```diff
@@ -1,32 +1,54 @@
 package de.fraunhofer.fokus.OpenMobileNetworkToolkit.cloudCity;
 
+import android.util.Log;
+
 import java.util.Locale;
+import java.util.concurrent.locks.ReentrantLock;
 
 import retrofit2.Retrofit;
 import retrofit2.converter.gson.GsonConverterFactory;
 
 public class NetworkClient {
+    private static final String TAG = "NetworkClient";
 
-    private static Retrofit retrofit;
+    private static volatile Retrofit retrofit;
+    private static volatile String currentClientBaseUrl;
+    private static ReentrantLock lock = new ReentrantLock();
+    private static final Object lockObject = new Object();
```

Comment:
1. the synchronized method is fine for synchronizing execution of this method alone.
2. the volatile variables ensure that, even if a member is accessed directly (e.g. by making it non-private via reflection) they will still have to be read and not be cached in the local thread cache (which is important when in use by multiple threads)
3. the single lock is still there.

The multiple synchronization techniques are all fighting different synchronization-related problems, and even then - this will not be targetted by multiple threads concurrently, but by just one place alone for now. 

However, we might as well make it proper from the start, rather than symptomatically fixing these issues as they creep up.

thoughts?